File: merge/merge_parallel_new.py
from parallel_utils import run_in_threads


def _merge_block(source, dest, left, mid, right):
    i, j, k = left, mid, left
    while i < mid and j < right:
        if source[i] <= source[j]:
            dest[k] = source[i]
            i += 1
        else:
            dest[k] = source[j]
            j += 1
        k += 1

    while i < mid:
        dest[k] = source[i]
        i += 1
        k += 1

    while j < right:
        dest[k] = source[j]
        j += 1
        k += 1


def parallel_merge_sort(arr, left, right):
    if left >= right:
        return arr

    data = arr[left : right + 1]
    n = len(data)
    buffer = [None] * n
    width = 1
    source = data
    dest = buffer

    while width < n:
        block_count = (n + 2 * width - 1) // (2 * width)

        def merge_blocks(blocks):
            for block in blocks:
                lo = block * 2 * width
                mid = min(lo + width, n)
                hi = min(lo + 2 * width, n)
                _merge_block(source, dest, lo, mid, hi)

        run_in_threads(block_count, merge_blocks)
        source, dest = dest, source
        width *= 2

    arr[left : right + 1] = source[:]
    return arr

File: merge/merge_parallel_new.cpp
#include <algorithm>
#include <vector>

namespace merge_parallel_detail {

void mergeBlock(const std::vector<int>& source, std::vector<int>& dest, int left, int mid, int right) {
    int i = left;
    int j = mid;
    int k = left;

    while (i < mid && j < right) {
        if (source[static_cast<std::size_t>(i)] <= source[static_cast<std::size_t>(j)]) {
            dest[static_cast<std::size_t>(k++)] = source[static_cast<std::size_t>(i++)];
        } else {
            dest[static_cast<std::size_t>(k++)] = source[static_cast<std::size_t>(j++)];
        }
    }

    while (i < mid) {
        dest[static_cast<std::size_t>(k++)] = source[static_cast<std::size_t>(i++)];
    }
    while (j < right) {
        dest[static_cast<std::size_t>(k++)] = source[static_cast<std::size_t>(j++)];
    }
}

}  // namespace merge_parallel_detail

void parallelMergeSort(std::vector<int>& arr, int left, int right) {
    if (left >= right) {
        return;
    }

    std::vector<int> source(arr.begin() + left, arr.begin() + right + 1);
    std::vector<int> dest(source.size());
    const int n = static_cast<int>(source.size());

    for (int width = 1; width < n; width *= 2) {
        const int block_count = (n + 2 * width - 1) / (2 * width);

#ifdef _OPENMP
#pragma omp parallel for schedule(static)
#endif
        for (int block = 0; block < block_count; ++block) {
            const int lo = block * 2 * width;
            const int mid = std::min(lo + width, n);
            const int hi = std::min(lo + 2 * width, n);
            merge_parallel_detail::mergeBlock(source, dest, lo, mid, hi);
        }

        source.swap(dest);
    }

    std::copy(source.begin(), source.end(), arr.begin() + left);
}
